### ISSUE TITLE
Better support for erase() with range-based iteration

### ISSUE THREAD

**Issue by jsyjr** [martinus/unordered_dense#43]
This a following from the porting exercise described in issue #42.

An issue I have encountered is that unordered_dense does handle erase within range-base iteration:
```
for (auto i : myUnorderedDense) {
    if (p(i)) {
        myUnorderedDense.erase(i);
    }
}
```
I do not have the standard's wording but [cppreference.com](https://en.cppreference.com/w/cpp/container/set/erase) states:

> References and iterators to the erased elements are invalidated. Other references and iterators are not affected.

I acknowledge that such broad invariants are impossible to satisfy.  Still, by exploiting the fact that these containers are **unordered**, the above code could be made to work.  As far as I can see, there is neither requirement, nor expectation, that iteration will occur in forward order over the values vector.

So, if instead `table::begin()` returned `m_values.rbegin()` and `table::end()` returned `m_values.rend()` then range-based iteration would work as expected.

**Comment by martinus** [comment]
Unfortunately iterating in reverse order doesn't help for this map. The problem is that whenever an element is erased, I have to replace that empty place with the last element in the vector. So no matter if you iterate front to back or back to front, there's going to be an iterator to an existing element that is going to be invalidated on erase.

**Comment by jsyjr** [comment]
I beg to differ.

The ends of the vector are not interchangeable.  The head is stable, the tail gets manipulated by erasure.  In the current implementation using forward iteration, erasing the current entry does two bad things:

* It moves the, as yet unvisited, tail value into a position that has already been visited.  That moved value will never get seen during iteration.
* The iterator controlling the for loop now points at destroyed storage.  Continuing the loop ventures into undefined behavior.

By contrast, if iteration proceeds from tail to head, then things work much better:

* Erasing the current value moves an already visited value into the current location.  This is harmless as iteration will never visit this current position again.  (Each subsequent iteration will visit a position ever closer to the stable head of the vector.)
* Since the iterator controlling for loop termination points to the stable head of the vector, continued iteration remains well defined.

**Comment by martinus** [comment]
> I beg to differ.

My concern was that iterators to elements that are shuffled around will be invalidated after the loop anyways, so there's not much gained from iterating in reverse order.

> That moved value will never get seen during iteration.

That only happens when you don't use the returned iterator of `erase()`. Something like this works:

```cpp
for (auto it = map.begin(); it != map.end();) {
    if (predicate(it)) {
        it = container.erase(it);
    } else {
        ++it;
    }
}
```

Also I've implemented `std::erase_if` which is most likely the best (fastest & simplest) option for this:

```cpp
std::erase_if(map, predicate);
```

I think the old form here has actually always been problematic:

```cpp
for (auto i : myUnorderedDense) {
    if (p(i)) {
        myUnorderedDense.erase(i);
    }
}
```

Because `erase(i)` invalidates `i`, but to get to the next iterator position the loop still has to do `++i`, which is not allowed any more after the call to erase()

**Comment by jsyjr** [comment]
Thanks for this helpful reply.

We have erase loops, similar to what I presented in my previous note, that operate successfully under our in-house robin hood implementation.  Within our code base, I have also seen code that looks like the first snippet in your most recent note.  All things being equal, I find the range-based formulation decidedly more readable.  Hence my effort to get your unordered_dense implementation to support it. 

Still, you make a very good point, that range-based iteration functions only because it advances an iterator that was invalidated by the preceding erase.  (The text I quoted mentioned that invalidation.  I failed to remember  the hidden ++ in the range-based for loop expansion.)

I have now looked at your std::erase_if.  I notice that it operates back to front for much the same reason that I was arguing.  I will switch our range-based erase loops to use erase_if.

Thanks again, /john
